src/operator/nn/pool.cuh

```diff
@@ -0,0 +1,826 @@
+          }
+        }
+      }
+    }
+    KERNEL_ASSIGN(out_data[index], req_type, max_val);
+  }
+}
+
+/*!
+ * \brief avg/sum pooling gpu kernel for 1-D images.
+ * Do not call this kernel directly. Use the interface pool().
+ */
+template <typename DType>
+__global__ void pool_sum_1d_gpu_kernel(const int nthreads, const DType* in_data, const int channels,
+                                       const int width, const int pooled_width, const int kernel_w,
+                                       const int stride_w, const int pad_w, OpReqType req_type,
+                                       DType* out_data, bool getAvg = false) {
+  CUDA_KERNEL_LOOP(index, nthreads) {
+	  const int pw = index % pooled_width;
+	  const int c = (index / pooled_width) % channels;
+	  const int n = index / pooled_width / channels;
+	  int wstart = pw * stride_w - pad_w;
+	  int wend = min(wstart + kernel_w, width + pad_w);
+	  const int pool_size = (getAvg? (wend - wstart) : 1);
+	  wstart = max(wstart, 0);
+	  wend = min(wend, width);
+	  DType sum = 0;
+	  const DType* out_slice =
+	 		in_data + (n * channels + c) * width;
+    for (int w = wstart; w < wend; ++w) {
+      sum += out_slice[w];
+    }
+	  KERNEL_ASSIGN(out_data[index], req_type, sum / pool_size);
+  }
+}
+
+/*!
+ * \brief avg/sum pooling gpu kernel for 2-D images.
+ * Do not call this kernel directly. Use the interface pool().
+ */
+template <typename DType>
+__global__ void pool_sum_2d_gpu_kernel(const int nthreads, const DType* in_data, const int channels,
+                                       const int height, const int width,
+                                       const int pooled_height, const int pooled_width,
+                                       const int kernel_h, const int kernel_w,
+                                       const int stride_h, const int stride_w,
+                                       const int pad_h, const int pad_w, OpReqType req_type,
+                                       DType* out_data, bool getAvg = false) {
+  CUDA_KERNEL_LOOP(index, nthreads) {
+	  const int pw = index % pooled_width;
+	  const int ph = (index / pooled_width) % pooled_height;
+	  const int c = (index / pooled_width / pooled_height) % channels;
+	  const int n = index / pooled_width / pooled_height / channels;
+	  int hstart = ph * stride_h - pad_h;
+	  int wstart = pw * stride_w - pad_w;
+	  int hend = min(hstart + kernel_h, height + pad_h);
+	  int wend = min(wstart + kernel_w, width + pad_w);
+	  const int pool_size = (getAvg? (hend - hstart) * (wend - wstart) : 1);
+	  hstart = max(hstart, 0);
+	  wstart = max(wstart, 0);
+	  hend = min(hend, height);
+	  wend = min(wend, width);
+	  DType sum = 0;
+	  const DType* out_slice =
+	 		in_data + (n * channels + c) * height * width;
+	  for (int h = hstart; h < hend; ++h) {
+		  for (int w = wstart; w < wend; ++w) {
+		    sum += out_slice[h * width + w];
+		  }
+	  }
+	  KERNEL_ASSIGN(out_data[index], req_type, sum / pool_size);
```

Comment:
forward doesn't need req_type support for now. Its almost always writeto.
add a check in pool and remove this